### PR TITLE
lightning: avoid retry write rows and record duplicate rows in tidb backend

### LINKED ISSUE #29996: lightning max-error record tripled when backend is tidb
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
1. Do lightning import for data with 2 type errors (e.g. type error), and check max-error records in lightning_task_info.type_error_v1

Test data:
```
[root@centos76_vm type_error]# tail max_error.t2.1.csv
21348226,"lGztBz","2021-11-17 00:00:00"
89250625,"ltDGh","2021-11-17 00:00:00"
9207504,"iEMNwuP","2021-11-17 00:00:00"
1,"b1","2021-11-14 00:00:00"
2,"b2","2021-11-14 00:00:00"
3,"b3","2021-11-14 00:00:00"
4,"b4","2021-11-14 00:00:00"
"xxx","b5","2021-11-15 00:00:00"
5,"b6","0"

[root@centos76_vm type_error]# cat max_error.t2-schema.sql
/*!40101 SET NAMES binary*/;
CREATE TABLE `t2` (
  `a` int(11) NOT NULL,
  `b` varchar(10) DEFAULT NULL,
  `c` datetime DEFAULT NULL,
  PRIMARY KEY (`a`) /*T![clustered_index] CLUSTERED */,
  UNIQUE KEY `uk_b` (`b`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;

```
lightning config is as below
```SQL
[root@centos76_vm type_error]# grep -v ^# lightning.toml
[lightning]
level = "info"
file = "tidb-lightning.log"
max-error = 10

[checkpoint]
driver = "mysql"

[tikv-importer]
backend = "tidb"
sorted-kv-dir = "/home/sorted-kv-dir"

[mydumper]
data-source-dir = "s3://nfs/tools_test_data/max_error/type_error?access-key=minioadmin&secret-access-key=minioadmin&endpoint=http%3a%2f%2fminio.pingcap.net%3a9000&force-path-style=true"
no-schema = false

filter = ['*.*', 'mysql.*', '!sys.*', '!INFORMATION_SCHEMA.*', '!PERFORMANCE_SCHEMA.*', '!METRICS_SCHEMA.*', '!INSPECTION_SCHEMA.*']

[mydumper.csv]
separator = ','
delimiter = '"'
header = true


[tidb]
host = "172.16.6.217"
port = 4000
user = "root"
status-port = 10080
sql-mode = "STRICT_TRANS_TABLES"

```

### 2. What did you expect to see? (Required)
lightning_task_info.type_error_v1 should record 2 rows with type error

### 3. What did you see instead (Required)
lightning_task_info.type_error_v1 recorded 6 rows with type error, the records are tripled.
This issue is NOT seem when backend is local.

```SQL
mysql> select * from lightning_task_info.type_error_v1;
+---------------------+----------------------------+------------------+--------------------+--------+-------------------------------------------------------------------+------------------------------------+
| task_id             | create_time                | table_name       | path               | offset | error                                                             | row_data                           |
+---------------------+----------------------------+------------------+--------------------+--------+-------------------------------------------------------------------+------------------------------------+
| 1637567621075919308 | 2021-11-22 15:53:42.370553 | `max_error`.`t2` | max_error.t2.1.csv |   3803 | Error 1366: Incorrect int value: 'xxx' for column 'a' at row 1    | ('xxx','b5','2021-11-15 00:00:00') |
| 1637567621075919308 | 2021-11-22 15:53:42.379626 | `max_error`.`t2` | max_error.t2.1.csv |   3836 | Error 1292: Incorrect datetime value: '0' for column 'c' at row 1 | ('5','b6','0')                     |
| 1637567621075919308 | 2021-11-22 15:53:42.507533 | `max_error`.`t2` | max_error.t2.1.csv |   3803 | Error 1366: Incorrect int value: 'xxx' for column 'a' at row 1    | ('xxx','b5','2021-11-15 00:00:00') |
| 1637567621075919308 | 2021-11-22 15:53:42.512061 | `max_error`.`t2` | max_error.t2.1.csv |   3836 | Error 1292: Incorrect datetime value: '0' for column 'c' at row 1 | ('5','b6','0')                     |
| 1637567621075919308 | 2021-11-22 15:53:42.640734 | `max_error`.`t2` | max_error.t2.1.csv |   3803 | Error 1366: Incorrect int value: 'xxx' for column 'a' at row 1    | ('xxx','b5','2021-11-15 00:00:00') |
| 1637567621075919308 | 2021-11-22 15:53:42.645454 | `max_error`.`t2` | max_error.t2.1.csv |   3836 | Error 1292: Incorrect datetime value: '0' for column 'c' at row 1 | ('5','b6','0')                     |
+---------------------+----------------------------+------------------+--------------------+--------+-------------------------------------------------------------------+------------------------------------+
6 rows in set (0.00 sec)

```

### 4. What is your TiDB version? (Required)
[root@centos76_vm type_error]# tiup tidb-lightning:v5.3.0 -V
Starting component `tidb-lightning`: /root/.tiup/components/tidb-lightning/v5.3.0/tidb-lightning -V
Release Version: v5.3.0
Git Commit Hash: 79e237d9d336e78f1933e38940f2e1703e383b1f
Git Branch: heads/refs/tags/v5.3.0
Go Version: go1.16.4
UTC Build Time: 2021-11-18 13:06:05
Race Enabled: false

### LINKED ISSUE #30185: data race when test in br/pkg/lightning/log
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

make race

### 2. What did you expect to see? (Required)

nothing error

### 3. What did you see instead (Required)

```
WARNING: DATA RACE
Write at 0x0000089ee198 by goroutine 51:
  github.com/pingcap/tidb/br/pkg/lightning/log.InitLogger()
      /go/tidb/br/pkg/lightning/log/log.go:113 +0x364
  github.com/pingcap/tidb/br/pkg/lightning/restore.(*chunkRestoreSuite).TestEncodeLoopIgnoreColumnsCSV()
      /go/tidb/br/pkg/lightning/restore/restore_test.go:1492 +0x91
  runtime.call16()
      /usr/local/go/src/runtime/asm_amd64.s:550 +0x3d
  reflect.Value.Call()
      /usr/local/go/src/reflect/value.go:337 +0xd8
  github.com/pingcap/check.(*suiteRunner).forkTest.func1()
      /home/prow/go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/check.go:850 +0xb3b
  github.com/pingcap/check.(*suiteRunner).forkCall.func1()
      /home/prow/go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/check.go:739 +0x11d

Previous read at 0x0000089ee198 by goroutine 104:
  github.com/pingcap/tidb/br/pkg/lightning/restore.(*Controller).buildRunPeriodicActionAndCancelFunc.func6()
      /go/tidb/br/pkg/lightning/restore/restore.go:1145 +0x52b

Goroutine 51 (running) created at:
  github.com/pingcap/check.(*suiteRunner).forkCall()
      /home/prow/go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/check.go:734 +0x4cc
  github.com/pingcap/check.(*suiteRunner).forkTest()
      /home/prow/go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/check.go:832 +0x1c4
  github.com/pingcap/check.(*suiteRunner).doRun()
      /home/prow/go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/check.go:666 +0x144
  github.com/pingcap/check.(*suiteRunner).run()
      /home/prow/go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/check.go:696 +0x144
  github.com/pingcap/check.Run()
      /home/prow/go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/run.go:150 +0x5a
  github.com/pingcap/check.RunAll()
      /home/prow/go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/run.go:113 +0xff6
  github.com/pingcap/check.TestingT()
      /home/prow/go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/run.go:99 +0x756
  github.com/pingcap/tidb/br/pkg/lightning/restore.TestTiDB()
      /go/tidb/br/pkg/lightning/restore/tidb_test.go:45 +0x38
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1193 +0x202

Goroutine 104 (finished) created at:
  github.com/pingcap/tidb/br/pkg/lightning/restore.(*Controller).restoreTables()
      /go/tidb/br/pkg/lightning/restore/restore.go:1441 +0x427
  github.com/pingcap/tidb/br/pkg/lightning/restore.(*tableRestoreSuite).TestTableRestoreMetrics()
      /go/tidb/br/pkg/lightning/restore/restore_test.go:1088 +0x187d
  runtime.call16()
      /usr/local/go/src/runtime/asm_amd64.s:550 +0x3d
  reflect.Value.Call()
      /usr/local/go/src/reflect/value.go:337 +0xd8
  github.com/pingcap/check.(*suiteRunner).forkTest.func1()
      /home/prow/go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/check.go:850 +0xb3b
  github.com/pingcap/check.(*suiteRunner).forkCall.func1()
      /home/prow/go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/check.go:739 +0x11d
```

### PR DESCRIPTION
**Pull request by glorv**
### What problem does this PR solve?

Issue Number: close #29996

close #30185 by [3332896](https://github.com/pingcap/tidb/pull/30179/commits/3332896df718f3c7305d75a40499cf2c3a26f86a)

Problem Summary:

### What is changed and how it works?
- When met error and the rows are recorded, should return instead retry next loop.
- Change the error row row record with "," separated string instead of the error message in `EncodeRowForRecord` since this function is called because the outer call of `Encode` fails.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
  - Manually test with in a local env, the error record won't duplicate now. And when there is a `column count mismatch` the row_data is the a string represent of the original input instead of the error.
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/backend/tidb/tidb_test.go [hunk 1]
```diff
@@ -408,7 +409,33 @@ func TestFetchRemoteTableModels_4_x_auto_random(t *testing.T) {
...
+	)
+	dataRows := encodeRowsTiDB(t, ignoreBackend, s.tbl)
+	ctx := context.Background()
+	engine, err := ignoreBackend.OpenEngine(ctx, &backend.EngineConfig{}, "`foo`.`bar`", 1)
+	require.NoError(t, err)
+	writer, err := engine.LocalWriter(ctx, nil)
+	require.NoError(t, err)
+	err = writer.WriteRows(ctx, []string{"a"}, dataRows)
+	require.Error(t, err)
+	require.Equal(t, utils.IsRetryableError(err), false, "err: %v", err)
```
Comment (gozssky) [inline br/pkg/lightning/backend/tidb/tidb_test.go:435]:
```suggestion
	require.False(t, utils.IsRetryableError(err), "err: %v", err)
```

**Review by gozssky — APPROVED** [review]

File: br/pkg/lightning/backend/tidb/tidb.go [hunk 2]
```diff
@@ -351,7 +351,20 @@ func EncodeRowForRecord(encTable table.Table, sqlMode mysql.SQLMode, row []types
...
+		var res strings.Builder
+		for i, d := range row {
+			if i > 0 {
+				res.WriteByte(',')
+			}
+			s, err := d.ToString()
+			if err != nil {
+				// ignore the field if there are any error
+				s = ""
+			}
```
Comment (kennytm) [inline br/pkg/lightning/backend/tidb/tidb.go:364]: 🤔.

The only case `ToString()` returns an error is when `d.Kind()` is unknown which shouldn't happen. I'd suggest using `types.DatumsToString()` directly here.

**Review by kennytm — COMMENTED** [review]
rest lgtm

**Review by kennytm — APPROVED** [review]

**Pull request merged**